pkg/storegateway/indexheader/lazy_binary_reader.go

```diff
@@ -230,125 +268,170 @@ func (r *LazyBinaryReader) EagerLoad() {
 
-	r.readerMx.Lock()
-	defer r.readerMx.Unlock()
-
-	// Ensure none else tried to load it in the meanwhile.
-	if r.reader != nil {
-		return nil
-	}
-	if r.readerErr != nil {
-		return r.readerErr
-	}
-
 	level.Debug(r.logger).Log("msg", "lazy loading index-header file", "path", r.filepath)
 	r.metrics.loadCount.Inc()
 	startTime := time.Now()
 
 	reader, err := r.readerFactory()
 	if err != nil {
 		r.metrics.loadFailedCount.Inc()
-		r.readerErr = err
-		return errors.Wrapf(err, "lazy load index-header file at %s", r.filepath)
+		return nil, errors.Wrapf(err, "lazy load index-header file at %s", r.filepath)
 	}
 
-	r.reader = reader
 	elapsed := time.Since(startTime)
 
 	level.Debug(r.logger).Log("msg", "lazy loaded index-header file", "path", r.filepath, "elapsed", elapsed)
 	r.metrics.loadDuration.Observe(elapsed.Seconds())
 
-	return nil
+	return reader, nil
 }
 
-// unloadIfIdleSince closes underlying BinaryReader if the reader is idle since given time (as unix nano). If idleSince is 0,
-// the check on the last usage is skipped. Calling this function on a already unloaded reader is a no-op.
-func (r *LazyBinaryReader) unloadIfIdleSince(ts int64) error {
-	r.readerMx.Lock()
-	defer r.readerMx.Unlock()
-
-	// Nothing to do if already unloaded.
-	if r.reader == nil {
-		return nil
-	}
-
-	// Do not unloadIfIdleSince if not idle.
-	if ts > 0 && r.usedAt.Load() > ts {
-		return errNotIdle
+func (r *LazyBinaryReader) waitAndCloseReader(req readerRequest) {
+	resp := <-req.response
+	if resp.reader != nil {
+		resp.inUse.Done()
 	}
+}
 
-	// Wait until all users finished using current reader.
-	r.readerInUse.Wait()
-
-	r.metrics.unloadCount.Inc()
-	if err := r.reader.Close(); err != nil {
-		r.metrics.unloadFailedCount.Inc()
-		return err
+// unloadIfIdleSince closes underlying BinaryReader if the reader is idle since given time (as unix nano). If idleSince is 0,
+// the check on the last usage is skipped. Calling this function on a already unloaded reader is a no-op.
+func (r *LazyBinaryReader) unloadIfIdleSince(tsNano int64) error {
+	req := unloadRequest{
+		// The channel is unbuffered because we will read the response. It should be buffered if we can give up before reading from it
+		response:       make(chan error),
+		idleSinceNanos: tsNano,
 	}
-
-	r.reader = nil
-	return nil
+	r.unloadReqC <- req
+	return <-req.response
 }
 
-// isIdleSince returns true if the reader is idle since given time (as unix nano).
-func (r *LazyBinaryReader) isIdleSince(ts int64) bool {
-	if r.usedAt.Load() > ts {
-		return false
+func (r *LazyBinaryReader) controlLoop() {
+	var loaded loadedReader
+	for {
+		select {
+		case readerReq := <-r.loadedReaderC:
+			if loaded.reader != nil {
+				loaded.inUse.Add(1)
+				r.usedAt.Store(time.Now().UnixNano())
+			}
+			readerReq.response <- loaded
+
+		case loadReq := <-r.loadReqC:
+			if loaded.reader == nil && loaded.err == nil {
+				loaded = loadedReader{}
+				loaded.reader, loaded.err = r.loadReader()
```

Comment:
It seems that once the error is stored, nothing cleans it, is that right? A transient error (disk is full), would prevent this block from being loaded ever again.

Reply:
i had the same doubt. I erred on the side of retaining existing behaviour. I don't see why we'd want to retain the error to be honest.

Reply:
I will take another look and just enable retries if there are no negative consequences.

Reply:
i think it makes sense. The risk is that we try to load a bad index header which consumes a lot of resources and eventually fails. I can't think of such a realistic case, so I did the change to retry loading it